client/service/src/task_manager/mod.rs

```diff
@@ -127,7 +124,20 @@ impl SpawnTaskHandle {
 			}
 		};
 
-		let join_handle = self.executor.spawn(Box::pin(future.in_current_span()), task_type);
+		let future: Pin<Box<dyn Future<Output = ()> + Send>> =
+			if let Some(telemetry_span) = self.telemetry_span.clone() {
+				// this will preserve the telemetry span and by extension all of its parents as they
+				// are linked together upon creation
+				// this is used by sc_telemetry::layer::TelemetryLayer
+				Box::pin(future.instrument(telemetry_span.span()))
+			} else {
+				// this will preserve the current span (if any) and by extension all of its parents
+				// as they are linked together upon creation
+				// this is used by sc_tracing::logging::prefix_logs_with
+				Box::pin(future.in_current_span())
```

Comment:
If I do the following:
```
tracing::info_span!("hello");
spawn_handle.spawn(something);
```

When there is a telemetry span, my custom spawn wouldn't be passed. While it would be passed when there isn't a telemetry span.

Now thinking about this. Why is the prefix span for example passed here implicitly and telemetry span needs to be passed explicitly?  Is the reason to not forget the telemetry span?

Reply:
- The prefix span is passed like a decorator and wraps the entire node initialization code. It's written in the user code.
- The telemetry span is created internally with the creation of the node parts. It's written in our code. To make it propagate we would need to `enter()` the span in the user code.

In a previous iteration of the big telemetry PR I did use a trick to enter the span without requiring the user to add it in their code but it has the inconvenient that the span is not exited easily. It was not really a problem because there was no other code.

So yeah we could use that trick again. If we do it, we can simplify the API by removing the telemetry span returned by `new_full_parts` and the user won't need to pass it around anymore. But it won't be possible for the user to explicitly enter the span anywhere they want. I suspect this can be useful for things like the collator in cumulus.

Reply:
Couldn't it be part of the configuration or similar, so that the user still has access to it? But most of it is done automatically for them?

Reply:
:thinking: good point... I didn't think about that but I can certainly add it to the configuration in the runner :bulb: you: :brain:

Reply:
Works nicely :ok_hand: